zerver/tests/test_auth_backends.py

```diff
@@ -1891,6 +1891,65 @@ def test_validate_state(self) -> None:
             self.assertIn('login', result.url)
             m.assert_called_with("Sign in with Apple failed: bad state token.")
 
+class AppleAuthBackendNativeFlowTest(AppleIdAuthBackendTest):
+    SIGNUP_URL = '/complete/apple/'
+
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         desktop_flow_otp: Optional[str]=None,
+                         is_signup: bool=False,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         alternative_start_url: Optional[str]=None,
+                         user_agent: Optional[str]=None,
+                         **extra_data: Any) -> HttpResponse:
+        url, headers = self.prepare_login_url_and_headers(
+            subdomain, mobile_flow_otp, desktop_flow_otp, is_signup, next,
+            multiuse_object_key, alternative_start_url=self.AUTH_FINISH_URL,
+            user_agent=user_agent,
+        )
+
+        params = {'native_flow': 'true'}
+
+        if not extra_data.get('skip_id_token'):
+            params['id_token'] = self.generate_id_token(account_data_dict)
+
+        if is_signup:
+            params['is_signup'] = '1'
+
+        if subdomain:
+            params['subdomain'] = subdomain
+
+        url += "&%s" % (urllib.parse.urlencode(params),)
```

Comment:
I think we should just override `prepare_login_url_and_headers` to do `super().prepare_login_url_and_headers` and then this logic above - because that's where the url construction naturally belongs.